codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,339 @@
+	protoTag, ok := lookupProtoTag(tag)
+	if !ok {
+		return "", 0, 0, errNoProtoTag
+	}
+	byCommaSplits := strings.Split(protoTag, ",")
+	if len(byCommaSplits) == 0 {
+		// Strange, how can a protobuf tag not be populated?
+		return "", 0, 0, errNoProtoTag
+	}
+	// By this point our field's struct tags are of the form:
+	//      Id int32 `protobuf:"varint,1,opt,name=id,proto3" json:"id,omitempty"`
+	wireType, protoFieldNumStr := byCommaSplits[0], byCommaSplits[1]
+	protoField, err := strconv.ParseUint(protoFieldNumStr, 10, 64)
+	if err != nil {
+		return "", 0, 0, fmt.Errorf("invalid protoFieldNum in %q", protoFieldNumStr)
+	}
+	var fieldName string
+	nameEqValue := byCommaSplits[3]
+	byEqualsSplit := strings.Split(nameEqValue, "=")
+	if len(byEqualsSplit) > 1 && byEqualsSplit[0] == "name" {
+		fieldName = byEqualsSplit[1]
+	}
+
+	wireTypeType, ok := wireTypeStrToWireType[wireType]
+	if !ok {
+		// Sanity check just in case of an unhandled type, but shouldn't occur.
+		panic(fmt.Sprintf("unhandled wireType %q", wireType))
+	}
+
+	return fieldName, protowire.Number(protoField), wireTypeType, nil
+}
+
+// CheckMismatchedProtoFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType, as well as mismatched
+// wireTypes.
+func CheckMismatchedProtoFields(b []byte, msg gogoproto.Message) error {
+	if msg == nil {
+		return errors.New("expecting a non-nil message")
+	}
+	rv := reflect.ValueOf(msg)
+	rt := rv.Type()
+
+	index, err := indexFieldsForType(rt)
+	if err != nil {
+		return err
+	}
+
+	protoNumToGoTypesIndex, err := protoNumToGoType(rt)
+	if err != nil && err != errNoProtoTag {
+		return err
+	}
+
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("invalid length")
+		}
+
+		twt, ok := index[tagNum]
+		switch {
+		case ok:
+			// Now assert that the wireTypes match.
+			if twt.wireType != wireType {
+				return &MismatchedWireType{
+					Type:         rt.String(),
+					TagNum:       tagNum,
+					GotWireType:  wireType,
+					WantWireType: twt.wireType,
+				}
+			}
+
+		default:
+			if tagNum&bit11NonCritical == 0 {
+				// The tag is non-critical, so report it.
+				return &UnexpectedField{
+					Type:     rt.String(),
+					TagNum:   tagNum,
+					WireType: wireType,
+				}
+			}
+		}
+
+		// Given that the wireTypes match, now traverse that field.
+		// Find the field whose protoNumber was just looked at.
+		// Skip over the 2 bytes that store fieldNumber and wireType bytes.
+		fieldBytes := b[2:n]
+		b = b[n:]
+
+		fieldGoType, ok := protoNumToGoTypesIndex[tagNum]
+		if !ok {
+			continue
+		}
+
+		if !fieldGoType.Implements(reflectGogoProtoDotMessage) {
+			// TODO: Perhaps try its pointer or a variant of it aka the opposite to check
+			// if that implements gogoproto.Message?
+			continue
```

Comment:
I've taken the hybrid approach of reading from Proto descriptors, but this case will ALWAYS need to be included, in case the underlying reflect.Type maybe needs a different form of indirection e.g. **type instead of what we expect *type, so unrelated to concerns.